pkg/gameserversets/allocation_overflow.go

```diff
@@ -0,0 +1,162 @@
+	c.baseLogger.Debug("Created!")
+	c.workerqueue = workerqueue.NewWorkerQueueWithRateLimiter(c.syncGameServerSet, c.baseLogger, logfields.GameServerSetKey, agones.GroupName+".GameServerSetController", workerqueue.FastRateLimiter(3*time.Second))
+	health.AddLivenessCheck("gameserverset-allocationoverflow-workerqueue", c.workerqueue.Healthy)
+
+	gsSetInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newGss := newObj.(*agonesv1.GameServerSet)
+
+			// Only process if there is an AllocationOverflow, and it has labels or annotations.
+			if newGss.Spec.AllocationOverflow == nil {
+				return
+			} else if len(newGss.Spec.AllocationOverflow.Labels) == 0 && len(newGss.Spec.AllocationOverflow.Annotations) == 0 {
+				return
+			}
+			if newGss.Status.AllocatedReplicas > newGss.Spec.Replicas {
+				c.workerqueue.Enqueue(newGss)
+			}
+		},
+	})
+
+	return c
+}
+
+// Run this controller. Will block until stop is closed.
+func (c *AllocationOverflowController) Run(ctx context.Context) error {
+	c.baseLogger.Debug("Wait for cache sync")
+	if !cache.WaitForCacheSync(ctx.Done(), c.gameServerSynced, c.gameServerSetSynced) {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	c.workerqueue.Run(ctx, 1)
+	return nil
+}
+
+// syncGameServerSet checks to see if there are overflow Allocated GameServers and applied the labels and/or
+// annotations to the requisite number of GameServers needed to alert the underlying system.
+func (c *AllocationOverflowController) syncGameServerSet(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// don't return an error, as we don't want this retried
+		runtime.HandleError(loggerForGameServerSetKey(c.baseLogger, key), errors.Wrapf(err, "invalid resource key"))
+		return nil
+	}
+
+	gsSet, err := c.gameServerSetLister.GameServerSets(namespace).Get(name)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			loggerForGameServerSetKey(c.baseLogger, key).Debug("GameServerSet is no longer available for syncing")
+			return nil
+		}
+		return errors.Wrapf(err, "error retrieving GameServerSet %s from namespace %s", name, namespace)
+	}
+
+	// just in case something changed, double check to avoid panics and/or sending work to the K8s API that we don't
+	// need to
+	if gsSet.Spec.AllocationOverflow == nil {
+		return nil
+	}
+	if gsSet.Status.AllocatedReplicas <= gsSet.Spec.Replicas {
+		return nil
+	}
+
+	overflow := gsSet.Status.AllocatedReplicas - gsSet.Spec.Replicas
+
+	list, err := ListGameServersByGameServerSetOwner(c.gameServerLister, gsSet)
+	if err != nil {
+		return err
+	}
+
+	matches, rest := gsSet.Spec.AllocationOverflow.CountMatches(list)
+	if matches >= overflow {
+		return nil
+	}
+
+	rest = sortGameServersByStrategy(gsSet.Spec.Scheduling, rest, c.counter.Counts())
```

Comment:
nit: shall we add some documentation about the sorting logic when picking up GameServers to add the overflow annotations/labels? Either in the design issue or somewhere on the website. I think it's kind of important for customer to be aware of, but right now it's not clear without looking into the code.